ls1intum/Artemis
# `Development`: Fix availability check for build job processing in local continuous integration

### Checklist
#### General


#### Server

### Motivation and Context


Sometimes, when multiple build jobs get added to the shared build job queue, it may happen that a single server instance adds more build jobs to its executor service than it has threads available. This is not critical as the build jobs will eventually get processed by the executor service, however it makes processing less efficient and could possibly lead to other issues.

### Description

This PR fixes the node availability checking. We check how many jobs are currently running on the node by using a counter. Then we check this against the maximum number of threads that can run on the node.

### Steps for Testing

**Actual change is hard to demonstrate without using specific tools**

Prerequisites:
- 1 Instructor

1. Trigger multiple build jobs (e.g. by creating two programming exercises)
2. Build results should be as expected (no build fail, all results eventually come in)

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
